app/src/pages/playground/usePlaygroundPrompt.tsx

```diff
@@ -0,0 +1,211 @@
+      definition: t.definition,
+    })),
+    // @TODO(apowell): Add description and invocationParameters
+  } satisfies Partial<PromptVersion>;
+};
+
+const getLatestPromptVersion = (
+  prompt: usePlaygroundPromptQuery$data["prompt"]
+) => {
+  if (!prompt) {
+    return null;
+  }
+  return prompt?.promptVersions?.edges?.[0]
+    ?.promptVersion as Mutable<PromptVersion> | null;
+};
+
+const query = graphql`
+  query usePlaygroundPromptQuery($promptId: GlobalID!) {
+    prompt: node(id: $promptId) {
+      ... on Prompt {
+        id
+        name
+        createdAt
+        description
+        promptVersions(first: 1) {
+          edges {
+            promptVersion: node {
+              id
+              description
+              modelName
+              modelProvider
+              invocationParameters
+              templateType
+              template {
+                __typename
+                ... on PromptChatTemplate {
+                  messages {
+                    ... on TextPromptMessage {
+                      content
+                      role
+                    }
+                  }
+                }
+              }
+              tools {
+                __typename
+                definition
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+export const usePlaygroundPromptReference = (instanceId: number) => {
+  const instances = usePlaygroundContext((state) => state.instances);
+  const instance = instances.find((instance) => instance.id === instanceId);
+  if (!instance) {
+    throw new Error(`Instance ${instanceId} not found`);
+  }
+  const promptId = instance.prompt?.id;
+  const [queryReference, loadQuery, disposeQuery] =
+    useQueryLoader<usePlaygroundPromptQuery>(query);
+
+  useEffect(() => {
+    if (promptId) {
+      loadQuery({ promptId: promptId });
+      return () => disposeQuery();
+    }
+  }, [disposeQuery, loadQuery, promptId]);
+
+  return queryReference;
+};
+
+export type PromptQueryReference = NonNullable<
+  ReturnType<typeof useQueryLoader<usePlaygroundPromptQuery>>[0]
+>;
+
+export const usePlaygroundPrompt = (
+  instanceId: number,
+  queryReference: PromptQueryReference
+) => {
+  const updateInstance = usePlaygroundContext((state) => state.updateInstance);
+  const instances = usePlaygroundContext((state) => state.instances);
+  const instance = instances.find((instance) => instance.id === instanceId);
+  if (!instance) {
+    throw new Error(`Instance ${instanceId} not found`);
+  }
+  const { prompt } = usePreloadedQuery(query, queryReference);
+
+  useEffect(() => {
+    const latestPromptVersion = getLatestPromptVersion(prompt);
+    if (latestPromptVersion && latestPromptVersion.templateType === "CHAT") {
+      const newInstance = promptVersionToInstance(latestPromptVersion);
+      updateInstance({
+        instanceId,
+        patch: {
+          ...newInstance,
+        },
+      });
+    }
+  }, [instanceId, prompt, updateInstance]);
```

Comment:
Haven't run this code but this scares me a tad that the instance is getting automatically set in a hook. It feels like we should be doing this outside of a hook at the point of selection of a prompt - not as a side effect.

Reply:
Like we just need:

1.  in the PromptPlaygroundInstance to pre-load the template and place into state.
2. on prompt change, load the prompt, push into state.

Both of those can be run either during page load or on a change from the UI (e.g. the selection of a promt.

I think you can avoid the effects.

Reply:
Removed in favor of imperative fetch callback